### PR TITLE
feat: add minorUnitToMajorUnitOffset to realtime price query/subscription

### PR DESCRIPTION
**Pull request by dolcalmi**
- ~currencyUnit now is an Enum (default is `MINOR` but adds the option to include/return `MAJOR` based on a param)~
- Add minorUnitToMajorUnitOffset to realtimePrice query and subscriptions

### CODE REVIEW

File: src/graphql/main/schema.graphql [hunk 1]
```diff
@@ -929,7 +934,8 @@ input PriceInput {
 
 interface PriceInterface {
   base: SafeInt!
-  currencyUnit: String!
+  currencyUnit: CurrencyUnit!
```
Comment (nicolasburtey) [inline src/graphql/main/schema.graphql:937]: is this not used in prod already?

Comment (nicolasburtey) [reply src/graphql/main/schema.graphql:937]: I don't quite get why we want to return Major or Minor?  Why no already return Minor?

File: src/graphql/main/schema.graphql [hunk 2]
```diff
@@ -260,6 +260,11 @@ type Currency {
   symbol: String!
 }
 
+enum CurrencyUnit {
+  MAYOR
```
Comment (nicolasburtey) [inline src/graphql/main/schema.graphql:264]:
```suggestion
  MAJOR
```

File: src/graphql/main/schema.graphql (hunk 1 above)
Comment (dolcalmi) [reply src/graphql/main/schema.graphql:937]: this only a change for realtimePrice, I asked @UncleSamtoshi about this field and is not used also I am changing it because does not make sense to use `${currency}CENTS` for all currencies (specially for currencies that dont use CENTS as minor unit).

File: src/domain/fiat/display-currency.ts [hunk 3]
```diff
@@ -25,5 +26,34 @@ export const majorToMinorUnit = ({
...
+    const { minimumFractionDigits } = formatter.resolvedOptions()
+    switch (minimumFractionDigits) {
+      case 0:
+        return MajorExponent.ZERO
+      case 1:
+        return MajorExponent.ONE
+      case 3:
+        return MajorExponent.THREE
+      default:
+        return MajorExponent.STANDARD
```
Comment (vindard) [inline src/domain/fiat/display-currency.ts:54]: Do we think it's worth it to do something like this in case something weird comes back from the library?:

```
case 2:
        return MajorExponent.STANDARD
default:
        recordExceptionInCurrentSpan({
            error: new UnexpectedMinimumFractionDigitsError(`${minimumFractionDigits}`),
            level: ErrorLevel.Critical,
        })
        return MajorExponent.STANDARD
```

File: src/graphql/main/schema.graphql [hunk 4]
```diff
@@ -929,7 +934,8 @@ input PriceInput {
 
 interface PriceInterface {
   base: SafeInt!
-  currencyUnit: String!
+  currencyUnit: CurrencyUnit!
+  minorUnitToMajorUnitOffset: Int!
```
Comment (vindard) [inline src/graphql/main/schema.graphql:938]: I like this name. It's longer than something like `currencyMinorUnitOffset` but it removes all ambiguity.

File: src/graphql/main/schema.graphql (hunk 1 above)
Comment (vindard) [reply src/graphql/main/schema.graphql:937]: I think the more important information here might be more which currency this is (`USD`/`EUR`/`CRC`) and less whether it's `MAJOR`/`MINOR`. I can't think of a case where we'd need `MAJOR` sometime soon... and if no-one else can either, then maybe we can deprecate this and re-introduce it when we need it again

File: src/graphql/root/subscription/my-updates.ts [hunk 5]
```diff
@@ -250,8 +271,8 @@ const MeSubscription = {
       const priceData = {
         timestamp: pricePerSat.timestamp,
         displayCurrency,
-        centsPerSat: usdMajorToMinorUnit(pricePerSat.price),
```
Comment (vindard) [inline src/graphql/root/subscription/my-updates.ts:253]: Did this used to be major unit coming out of `pricePerSat` before?

**Review by vindard — COMMENTED** [review]
Few questions/comments

File: src/domain/fiat/display-currency.ts (hunk 3 above)
Comment (dolcalmi) [reply src/domain/fiat/display-currency.ts:54]: :thinking: domain layer should not depend on service layer...

File: src/graphql/root/subscription/my-updates.ts (hunk 5 above)
Comment (dolcalmi) [reply src/graphql/root/subscription/my-updates.ts:253]: price service returns price per sat in major unit.

File: src/graphql/main/schema.graphql (hunk 1 above)
Comment (dolcalmi) [reply src/graphql/main/schema.graphql:937]: Add deprecation reason to currency unit https://github.com/GaloyMoney/galoy/pull/2414/commits/f48c847ae9cdff7666dd0d079aee0717b5b5633b

File: src/domain/fiat/display-currency.ts (hunk 3 above)
Comment (bodymindarts) [reply src/domain/fiat/display-currency.ts:54]: Domain layer shouldn't depend on external dependencies at all. If it does (eg as a wrapper like with the `domain/bitcoin` module) we need to deal with its possible return values explicitly to make it safe for other parts of the code to use.

Even if we don't want to return an exception we should at least add some attribute to make it visible that the library we are using isn't working as expected.

File: src/domain/fiat/display-currency.ts [hunk 6]
```diff
@@ -25,5 +26,34 @@ export const majorToMinorUnit = ({
...
+    }
+  } catch {
+    return MajorExponent.STANDARD
```
Comment (bodymindarts) [inline src/domain/fiat/display-currency.ts:57]: Silent catching is not good! Either this is a -we want it to crash- situation (ie remove try/catch) or we should be explicit about what we are catching and at least add it to the trace.

File: src/graphql/main/schema.graphql [hunk 7]
```diff
@@ -938,7 +945,9 @@ Price of 1 sat in base/offset. To calculate, use: `base / 10^offset`
 """
 type PriceOfOneSat implements PriceInterface {
   base: SafeInt!
-  currencyUnit: String!
+  currencyUnit: CurrencyUnit!
+    @deprecated(reason: "Deprecated in favor of minorUnitToMajorUnitOffset")
```
Comment (bodymindarts) [inline src/graphql/main/schema.graphql:949]: Which line is the `@deprecated` applying to? The `currencyUnit`?
Why are we changing the type if we are deprecating it?

File: src/graphql/types/scalar/currency-unit.ts [hunk 8]
```diff
@@ -0,0 +1,11 @@
+import { GT } from "@graphql/index"
+
+const CurrencyUnit = GT.Enum({
```
Comment (bodymindarts) [inline src/graphql/types/scalar/currency-unit.ts:3]: why is this added if its also deprecated instantly?

File: src/graphql/main/schema.graphql (hunk 7 above)
Comment (dolcalmi) [reply src/graphql/main/schema.graphql:949]: now `${currency}CENTS` does not make sense... the deprecation was requested in the last display currency meeting with mobile devs

File: src/graphql/types/scalar/currency-unit.ts (hunk 8 above)
Comment (dolcalmi) [reply src/graphql/types/scalar/currency-unit.ts:3]: https://github.com/GaloyMoney/galoy/pull/2414#discussion_r1135669342

File: src/domain/fiat/display-currency.ts (hunk 6 above)
Comment (dolcalmi) [reply src/domain/fiat/display-currency.ts:57]: so tracing service is an exception? I can add a comment but this is necessary for alias currencies (we dont want to crash to get the major exponent)

**Comment by dolcalmi** [comment]
closed in favor of https://github.com/GaloyMoney/galoy/pull/2439/
